dask/order.py

```diff
@@ -115,786 +106,427 @@ def order(
+                            if (
+                                len(known_runnable_paths[current])
+                                >= num_needed[current]
+                            ):
+                                pruned_branches: deque[list[Key]] = deque()
+                                for path in known_runnable_paths.pop(current):
+                                    if path[-2] not in result:
+                                        pruned_branches.append(path)
+                                if len(pruned_branches) < num_needed[current]:
+                                    known_runnable_paths[current] = list(
+                                        pruned_branches
+                                    )
+                                else:
+                                    if validate:
+                                        nodes_in_branches = set()
+                                        for b in pruned_branches:
+                                            nodes_in_branches.update(b)
+                                        cond = not (
+                                            dependencies[current]  # type: ignore
+                                            - set(result)
+                                            - nodes_in_branches
+                                        )
+                                        assert cond
+                                    while pruned_branches:
+                                        path = pruned_branches.popleft()
+                                        for k in path:
+                                            if num_needed[k]:
+                                                pruned_branches.append(path)
+                                                break
+                                            add_to_result(k)
+                        else:
+                            if (
+                                len(dependencies[current]) > 1  # type: ignore
+                                and num_needed[current] <= 1
+                            ):
+                                for k in path:
+                                    add_to_result(k)
+                            else:
+                                known_runnable_paths[current] = [path]
+                        break
 
-    result: dict[Key, int] = {}
-    i = 0
+    def pick_strategy() -> bool:
+        # Note: We're trying to be smart here by picking a strategy on how to
+        # determine the critical path. This is not always clear and we may want
+        # to consider just calculating both orderings and picking the one with
+        # less pressure. The only concern to this would be performance but at
+        # time of writing, the most expensive part of ordering is the prep work
+        # (various metrics, connected roots, etc.) which can be reused for
+        # multiple orderings.
+        size = 0
+        if not abs(len(root_nodes) - len(leaf_nodes)) / len(root_nodes) > 0.8:
```

Comment:
I'm not sure if this is intended, but the ratios of root and leaf nodes this algorithm considers are very uneven:

For `n_root_nodes := 10`: `abs(10 - n_leaf_nodes) / 10 > 0.8` if `n_leaf_nodes < 2`, i.e., a ratio of more than `5:1`. 
Also, `abs(10 - n_leaf_nodes) / 10 > 0.8` if `n_leaf_nodes > 18`, a ratio of less than `5:9` (or simplifying a bit `1:2`).

To make this more even, we may want to look at `abs(len(root_nodes) - len(leaf_nodes)) / {max|min}(len(root_nodes), len(leaf_nodes))`.

Reply:
This heuristic is a little hand wavy and I can not guarantee that it is correct. This asymmetry is actually accidental but when I tried modifying it I get various test failures that are rightfully false. Just from looking at a couple of examples, it appears as if this kind of asymmetry is beneficial but I believe we do not have sufficient test cases to say for sure.

I'll keep it as is for now